py/torch_tensorrt/dynamo/conversion/impl/linear.py

```diff
@@ -52,3 +56,107 @@ def linear(
         out = impl.elementwise.add(ctx, target, source_ir, name, out, bias)
 
     return out
+
+
+def cdist_forward(
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    x1: TRTTensor,
+    x2: TRTTensor,
+    p: float,
+    compute_mode: Optional[int] = None,
```

Comment:
It seems that `compute_mode` is not used below. Is there any difficulty for this? Ideally, we should mimic all pytorch's behaviors (i.e., receiving the same args and then returning the same outputs as pytorch, based on the [schema](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L4406C60-L4406C72)). 
If `compute_mode` cannot be supported for some reasons, we need to use `capability_validator` to make sure this kind of cases will not be converted, instead, falling back to pytorch.

Reply:
`compute_mode` does not influence the output values. It is designed for optimizing computational efficiency in specific situations when calculating the Euclidean distance (p=2) for large-sized inputs. 

I have modified to clarify:
1) A note has been added that `compute_mode` doesn't alter the computational path in our implementation.
2) A default value of 0 is used for `compute_mode` when it's unspecified, maintaining behavior consistency with Pytorch.
3) I have included a warning for situations where an optimized path for p=2 might be expected but isn't utilized due to our unified approach (element-wise).

Reply:
Thanks for the explanation! I'm still thinking if using `matmul` will be faster for the previous two scenarios? Considering that we have out-of-the-box [matmul](https://github.com/pytorch/TensorRT/blob/main/py/torch_tensorrt/dynamo/conversion/impl/matmul.py#L13), if possible, could you compare their time costs?

Reply:
Thank you for your valuable comment. In this update, I've implemented a matrix multiplication-based approach to compute the distance for `p=2`. During my tests on my local PC with inputs `x1=(150,100,50,50)` and `x2=(150,100,30,50)`, I observed that the TRT run time improved by approximately `6.3 times`. Here are the relevant logs from those tests:

```
INFO:harness:FX graph= graph():
    %x1 : [num_users=1] = placeholder[target=x1]
    %x2 : [num_users=1] = placeholder[target=x2]
    %_cdist_forward_default : [num_users=1] = call_function[target=torch.ops.aten._cdist_forward.default](args = (%x1, %x2, 2, 0), kwargs = {})
    return _cdist_forward_default
===============based on matmul : ==================
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.007999
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:255: DeprecationWarning: Use build_serialized_network instead.
  engine = self.builder.build_engine(self.ctx.net, builder_config)
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:33.710717
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 244801024 bytes of Memory
INFO:harness:Interpreter run time(s): 33.719337898997765
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:63: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.input_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:66: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.output_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:88: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:93: DeprecationWarning: Use get_tensor_shape instead.
  tuple(self.engine.get_binding_shape(idx))
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:98: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:247: DeprecationWarning: Use set_input_shape instead.
  self.context.set_binding_shape(
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:262: DeprecationWarning: Use get_tensor_shape instead.
  shape = tuple(self.context.get_binding_shape(idx))
INFO:harness:TRT run time(s)= 0.0026535038948059084
.INFO:harness:FX graph= graph():
    %x1 : [num_users=1] = placeholder[target=x1]
    %x2 : [num_users=1] = placeholder[target=x2]
    %_cdist_forward_default : [num_users=1] = call_function[target=torch.ops.aten._cdist_forward.default](args = (%x1, %x2, 2, 1), kwargs = {})
    return _cdist_forward_default
===============based on elementwise pow for diff : ==================
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.003120
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:255: DeprecationWarning: Use build_serialized_network instead.
  engine = self.builder.build_engine(self.ctx.net, builder_config)
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:43.902498
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 4590000640 bytes of Memory
INFO:harness:Interpreter run time(s): 43.90606521600421
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:63: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.input_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:66: DeprecationWarning: Use get_tensor_name instead.
  self.engine.get_binding_index(name) for name in self.output_names
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:88: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:93: DeprecationWarning: Use get_tensor_shape instead.
  tuple(self.engine.get_binding_shape(idx))
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:98: DeprecationWarning: Use get_tensor_dtype instead.
  self.engine.get_binding_dtype(idx), Frameworks.TORCH
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:247: DeprecationWarning: Use set_input_shape instead.
  self.context.set_binding_shape(
/home/hoonkyungc/workspace/TorchTRT/TensorRT/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:262: DeprecationWarning: Use get_tensor_shape instead.
  shape = tuple(self.context.get_binding_shape(idx))
INFO:harness:TRT run time(s)= 0.016830976486206056
.
----------------------------------------------------------------------
```

Reply:
It seems `matmul` runs faster in both compilation stage and running stage, and save a lot of memory.

Reply:
You're right. While always using matrix multiplication might seem more efficient considering TensorRT's capabilities, I've aligned the implementation with the compute mode of the ATen operations for consistency. Thank you for pointing this out!